### DISCUSSION TITLE
A path forward to enable comments?

### DISCUSSION THREAD

**Discussion by ymshulman** [roam-garden/i#20] [Feature requests & Ideas]
TL;DR: A backend that would allow building a comments feature into Roam.Garden, and integrate into our private graphs. Loom video [here](https://www.loom.com/share/e4084a58359649d19d5c8f48d63cb450), if you wish!
---
The ability to allow others to add comments to our Roam Garden graphs is something I imagine quite a few of us would love. 

The challenge, of course, is how to add them! Which needs to include a way for us to see the comments later, and import them back into our own graphs. Part of the challenge is the backend - where to store the metadata. Vlad had a conversation on Twitter about whether using hypothes.is (a web annotating tool) was an option, but it seems it is not (see [here](https://twitter.com/VladyslavSitalo/status/1342280453068267520?s=20)).

I had an idea, prompted by something Brandon Toner is doing with the shared Roam-collective graph. I posted about it on Twitter (see thread [here](https://twitter.com/ymshulman/status/1364364700901851136?s=20)), but wanted to open that discussion here, and expand on some parts.

The vision here is for a comment bubble on Roam.Garden, just like the one on Google Docs. Users open your graph, see a comment icon when they hover each block. The first time they click to comment it will prompt them for their name, and optionally prompt them for a team or organization name. They can comment, and can see comments from (depending on a setting you as the graph owner set) all other users, others on their team/in their organization, or only themselves. 

_(Seeing comments for others on their team is useful if you share your graph with a few collaborators and want to get feedback from a few of them.)_
![](https://firebasestorage.googleapis.com/v0/b/firescript-577a2.appspot.com/o/imgs%2Fapp%2FYmshulman%2FbiRNopSIVO.png?alt=media&token=f574ced6-b3a2-4ef6-a9d2-27337e167e76)
Now, how can we make this work? Ideally, where you can get those comments back in your graph? This is where Brandon Toner's work gave me an idea. Let RoamGarden create a new page of its own - that seems safe! Doesn't mess with your data, perfect. Let's call that page [[Roam.Garden.comments.2021]]. When a user clicks that comment icon and adds a comment, the comment gets added to this page - _under a nested hierarchy of the date, their team and their name_ (separate page mentions), **and then a block reference of the block they are commenting on**. 

Here's what happens! You have comments that are linked to specific blocks, all coming from a single page, that each mention the date, name, and maybe a team name. Roam.Garden can watch that single page for block references, and render them specially so they appear to a user who is in your Roam.Garden (if they are allowed to see those comments - either cause you've said everyone can see all comments, or because they or someone on their team created that comment.)

OK, we've created a way for Roam.Garden to have a backend of where to store comments and how to link them to blocks. It gets better!

You want those comments in your own graph, really. I mean - Roam.Garden is for other people, but you work in your graph! No problem. Import that single page [[Roam.Garden.comments.2021]] into your own graph. Here's what you get! Anywhere people have made a comment, you can click the number to the right of the block and see their comment, along with their name. Additionally, these comments will show up in the backlinks on your daily page - so you can see a quick overview of recent comments right on your daily page!

If Roam.Garden ever gets a javascript integration, so we can update our Roam.Gardens without the .json export dance, then that same integration can pull in the latest version of the [[Roam.Garden.comments.2021]] page. Boom! All comments integrated into your graph. Maybe someone can even come up with some CSS to style the number to the right of your blocks differently if some block references live on the [[Roam.Garden.comments.2021]] page. 

To be sure - I understand that Vlad would still need to build a system where Roam.Garden creates these comments popups, puts any comments in a specially formatted way into the [[Roam.Garden.comments.2021]] page, and constantly watches that page to render content from there as comments. But this implementation would provide the backend for that, and - critically - give graph owners a way to import these comments back into their own graphs, where they want them. 

![](https://firebasestorage.googleapis.com/v0/b/firescript-577a2.appspot.com/o/imgs%2Fapp%2FYmshulman%2FosRZrOjQ2S.png?alt=media&token=f089171c-079f-4207-bcf9-c00fe774f0de)

I hope this was clear! I love the ability to share pieces of my Garden with others, but allowing them to give me feedback on my work is a huge part of the value proposition of this Garden.

**Comment by Stvad** [comment]
Thanks a lot for such a detailed proposal!  

An ability to have social feedback on the content & being able to import it back to my graph is something that I'd really like to exist in Roam Garden!  

There a lot of things that I'd need to build before this proposal can come to fruition though. The biggest challenge is that any given instance of Roam Garden is currently a static site with no backend at all. And I wan to keep it that way for the foreseeable future. It should be possible to use the Hypothes.is API as a backbone for this even if the existing client won't work that well by itself.

So in a way this can be a custom integration with Hypothes.is + a separate roam/js plugin that would be able to pull comments from it for a given garden & store them in the format you propose

**Reply by ymshulman** [reply]
I have to admit that the coding part of this is all way beyond me - I suspect I'm even using the term "backend" wrong. I was guessing that all comments could be stored in a page as I described, and didn't think anything of Roam.Garden being a "static" site - figured it would just store one more page to enable comments, and use a combination of technologies similar to SmartBlocks to create comments, and a special instance of something like LivePreview to see comments alongside each page. 

But if there are other ways to make this happen, and the format is useful for how to get comments back into our graphs, then I'm glad this thought process may have been useful!
